tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala

```diff
@@ -0,0 +1,82 @@
+
+import java.io.ByteArrayOutputStream
+import java.nio.charset.StandardCharsets
+import java.util.concurrent.atomic.AtomicInteger
+import scala.jdk.OptionConverters._
+
+class EncryptedEmailContentFactory(encrypter: Encrypter) {
+
+  def encrypt(clearEmailContent: ClearEmailContent, messageId: MessageId): EncryptedEmailContent =
+    EncryptedEmailContent(encryptedPreview = encryptPreview(clearEmailContent.preview),
+      encryptedHtml = encrypt(clearEmailContent.html),
+      hasAttachment = clearEmailContent.hasAttachment,
+      encryptedAttachmentMetadata = encryptAttachmentMetadata(clearEmailContent.attachments, messageId),
+      encryptedAttachmentContents = encryptAttachmentContent(clearEmailContent.attachments))
+
+  def encryptPreview(preview: Preview): String =
+    encrypt(preview.getValue)
+
+  def encryptAttachmentMetadata(parsedAttachments: List[ParsedAttachment], messageId: MessageId): Option[String] = {
+    if (parsedAttachments.isEmpty) return None
+    Some(encrypt(AttachmentMetaDataSerializer
+      .serializeList(AttachmentMetadata.fromJava(parsedAttachments, messageId))
+      .toString()))
+  }
+
+  def encryptAttachmentContent(parsedAttachments: List[ParsedAttachment]): List[String] = {
+    if (parsedAttachments.isEmpty) return List.empty
+    parsedAttachments.map(parsedAttachment => encrypt(parsedAttachment.getContent))
+  }
+
+  def encrypt(byteSource: ByteSource): String = {
+    val stream: ByteArrayOutputStream = new ByteArrayOutputStream
+    encrypter.encrypt(byteSource, stream)
+    new String(stream.toByteArray, StandardCharsets.UTF_8)
+  }
+
+  def encrypt(value: String): String = {
+    val stream: ByteArrayOutputStream = new ByteArrayOutputStream
+    encrypter.encrypt(ByteSource.wrap(value.getBytes(StandardCharsets.UTF_8)), stream)
+    new String(stream.toByteArray, StandardCharsets.UTF_8)
+  }
+}
+
+case class EncryptedEmailContent(encryptedPreview: String,
+                                 encryptedHtml: String,
+                                 hasAttachment: Boolean,
+                                 encryptedAttachmentMetadata: Option[String],
+                                 encryptedAttachmentContents: List[String])
+
+object AttachmentMetadata {
+  def fromJava(parsedAttachment: ParsedAttachment, position: Int, messageId: MessageId): AttachmentMetadata =
+    AttachmentMetadata(
+      position = position,
+      blobId = s"encryptedAttachment_${messageId.serialize()}_$position",
+      name = parsedAttachment.getName.toScala,
+      contentType = parsedAttachment.getContentType.asString(),
+      cid = parsedAttachment.getCid
+        .toScala
+        .map(cid => cid.getValue),
+      isLine = parsedAttachment.isInline,
+      size = parsedAttachment.getContent.size())
+
+  def fromJava(listParsedAttachment: List[ParsedAttachment], messageId: MessageId): List[AttachmentMetadata] = {
+    val position: AtomicInteger = new AtomicInteger(0)
```

Comment:
I don't feel sure about this way (how to get `position` property)

Reply:
I think position is really what it says in the ticket. Like the first attachment will have position 0, second position 1, etc

Which logically should allow you to match with the correct encrypted attachment content.

Like the attachment with position 0 in its metadata, you access its content with `encryptedAttachmentContents[0]`, etc

Reply:
So, totally, we need to add one more property into `ParsedAttachment` (position), and need to refactor a `MessageParser`. Am I right?

Reply:
Nope, check this comment: https://github.com/linagora/tmail-backend/pull/175#discussion_r657613244 :)